### ISSUE TITLE
seastar build error due to a possible compiler bug

### ISSUE THREAD

**Issue by psankar** [scylladb/seastar#441]
I run into some g++ internal error compiling the HEAD (`71be10fb807e489d16a316674a26e79cc2c88ff7`) sources.

I am doing the Docker container based build on a Mac OS X machine (if it matters).

```
➜  seastar git:(master) seabuild ninja-build
[1/380] CXX build/debug/tests/thread_context_switch.o
FAILED: build/debug/tests/thread_context_switch.o
g++ -MD -MT build/debug/tests/thread_context_switch.o -MF build/debug/tests/thread_context_switch.o.d -fsanitize=address -fsanitize=leak -fsanitize=undefined -fno-sanitize=vptr -DSEASTAR_ASAN_ENABLED -O0 -DSEASTAR_DEBUG -DSEASTAR_DEBUG_SHARED_PTR -DSEASTAR_DEFAULT_ALLOCATOR -DSEASTAR_THREAD_STACK_GUARDS -DSEASTAR_NO_EXCEPTION_HACK -DSEASTAR_SHUFFLE_TASK_QUEUE -I/seastar/build/debug/gen -I/seastar/build/debug/c-ares -std=gnu++17 -g  -Wall -Werror -Wno-error=deprecated-declarations -fvisibility=hidden -Wno-attributes -pthread -I/seastar -U_FORTIFY_SOURCE  -I/seastar/fmt -DBOOST_TEST_DYN_LINK -Wno-overloaded-virtual -Wno-maybe-uninitialized -Wno-error=cpp -Wno-stringop-overflow -DSEASTAR_HAVE_HWLOC -DSEASTAR_HAVE_NUMA -DSEASTAR_HAS_MEMBARRIER -DSEASTAR_HAVE_LZ4_COMPRESS_DEFAULT -DSEASTAR_HAVE_ASAN_FIBER_SUPPORT -c -o build/debug/tests/thread_context_switch.o tests/thread_context_switch.cc
g++: internal compiler error: Killed (program cc1plus)
Please submit a full bug report,
with preprocessed source if appropriate.
See <http://bugzilla.redhat.com/bugzilla> for instructions.
[2/380] make -C build/debug/c-ares CC=gcc
make: Entering directory '/seastar/build/debug/c-ares'
make[1]: Entering directory '/seastar/build/debug/c-ares'
make[2]: Entering directory '/seastar/build/debug/c-ares'
Scanning dependencies of target c-ares
make[2]: Leaving directory '/seastar/build/debug/c-ares'
make[2]: Entering directory '/seastar/build/debug/c-ares'
```

**Comment by gleb-cloudius** [comment]
This looks like you ran out of memory.

--
			Gleb.

**Comment by psankar** [comment]
I re-ran the tests and had another crash, but this time at a different file. I was watching the memory growth for leaks from a top session in parallel. Strangely, there was no difference in the VIRT memory size and it remained stable at 6522 M , where as the resident memory changed from 2016M to just about 2044M after the above g++ error was reported. The build continued to run even after the above error message and did not get OOM-killed.

I suspect that the above is not due to memory pressure, but could actually be a threading issue. I've no proof for that though.

I am running this on a Macbook pro with 16 GB and RAM and there should be no memory pressure for this (unless there are restrictions via backdoor like ulimit, which is not very probable).

Also, the CPU was constantly at 200-300% and I actually wanted to run a non-parallel build to rule out any threading issue, but before all these, are there any recommendations for the building of these sources ? Do we have any minimum requirements like 32 GB RAM and 8 cores at least, etc. ?

Is there anything else I could pass (debug, compiler, build flags, etc.) to provide more context when builds crash ?

**Comment by nyh** [comment]
@psankar this is indeed most likely to be the out-of-memory killer, as @gleb-cloudius suggested.
Please try to compile with "ninja -j1" - it will only do one compilation in parallel (not number of cpus + 2) and use much less memory.
Note your compiler was *killed* - it didn't die of "natural" causes like a crash. So something must have done it.
Compiling Seastar applications requires huge amounts of memory - please allow about 4 GB per core. If you have too many cores and too little memory, pass the "ninja -n <number>" option to do less work in parallel even though you have many cores.

**Comment by psankar** [comment]
> Note your compiler was killed - it didn't die of "natural" causes like a crash. So something must have done it.

Got it :) What I meant was the entire compilation process did not get killed but only one file compilation in that failed and the build process proceeded after that. I realised that each file build is forked as a subprocess and so, only one of these build processes was killed and the overall compilation did not get interrupted fully. Thanks for the clarity.

I just started a -j1 and found that there is a kernel_task that is consuming 75G of VIRT memory with the name `kernel_task`. It is unlikely that this is due to the build tools and I wonder if the anti-virus that I am running is doing something bad while the build is going on.

I believe that this issue can be closed if no one else is facing this. I can track this offline too. Thanks.
